src/composer/compiler.rs

```diff
@@ -38,34 +54,266 @@ impl Compiler {
         pp: &PublicParameters,
         label: &[u8],
         circuit: &C,
-    ) -> Result<(Prover<C>, Verifier<C>), Error>
+    ) -> Result<(Prover, Verifier), Error>
     where
         C: Circuit,
     {
-        let max_size = (pp.commit_key.powers_of_g.len() - 1) >> 1;
-        let mut prover = Builder::initialized(max_size);
+        let max_size = Self::max_size(pp);
+        let mut builder = Builder::initialized(max_size);
+
+        circuit.circuit(&mut builder)?;
+
+        Self::compile_with_builder(pp, label, &builder)
+    }
 
-        circuit.circuit(&mut prover)?;
+    /// Compile a circuit and compress its prover and verifier into the provided
+    /// buffer, extending its bytes.
+    ///
+    /// Returns the amount of bytes written to `buffer`.
+    ///
+    /// [Self::decompress] is used to return the prover/verifier.
+    pub fn compress<C, B>(
+        pp: &PublicParameters,
+        buffer: &mut B,
+    ) -> Result<usize, Error>
+    where
+        C: Circuit,
+        B: Extend<u8>,
+    {
+        let max_size = Self::max_size(pp);
+        let mut builder = Builder::initialized(max_size);
+        let compressor = ScalarCompressor::default();
+
+        C::default().circuit(&mut builder)?;
+        let mut n = 0;
+
+        let witnesses = (builder.witnesses.len() as u64).to_le_bytes();
+        n += witnesses.len();
+        buffer.extend(witnesses);
+
+        let constraints = (builder.constraints.len() as u64).to_le_bytes();
+        n += constraints.len();
+        buffer.extend(constraints);
```

Comment:
Would be nice to add comment like:
// Store the amount of witnesses/constraint in the first/next 8 bytes

Reply:
Added on https://github.com/dusk-network/plonk/pull/752/commits/602a0c9ca782fa1e45451a65ac458276e6b1834e